### PR TITLE
[Improvement](top-n) support runtime update topn filter

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes
support runtime update topn filter
## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/olap/accept_null_predicate.h [hunk 1]
```diff
@@ -149,19 +151,10 @@ class AcceptNullPredicate : public ColumnPredicate {
...
     std::string get_search_str() const override { return _nested->get_search_str(); }
 
-    std::string debug_string() const override {
-        return "passnull predicate for " + _nested->debug_string();
-    }
-
-    /// Some predicates need to be cloned for each segment.
-    bool need_to_clone() const override { return _nested->need_to_clone(); }
-
     void clone(ColumnPredicate** to) const override {
```
Comment (HappenLee) [inline be/src/olap/accept_null_predicate.h:154]: here delete the need to clone code，why not delete `clone` interface also

File: be/src/olap/shared_predicate.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+class SharedPredicate : public ColumnPredicate {
+    ENABLE_FACTORY_CREATOR(SharedPredicate);
+
+public:
+    SharedPredicate(uint32_t column_id) : ColumnPredicate(column_id) {}
+
+    PredicateType type() const override {
+        std::shared_lock<std::shared_mutex> lock(_mtx);
+        if (!_nested) {
```
Comment (HappenLee) [inline be/src/olap/shared_predicate.h:40]: better DCHECK here

File: be/src/olap/shared_predicate.h [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            // topn filter is le or ge
+            return PredicateType::LE;
+        }
+        return _nested->type();
+    }
+
+    void set_nested(ColumnPredicate* nested) {
+        std::unique_lock<std::shared_mutex> lock(_mtx);
+        _nested.reset(nested);
```
Comment (HappenLee) [inline be/src/olap/shared_predicate.h:49]: why not init `_nested` in contructor， better and safer than `set_nested`

**Comment by BiteTheDDDDt** [comment]
run buildall

File: be/src/runtime/runtime_predicate.h [hunk 4]
```diff
@@ -52,25 +53,32 @@ class RuntimePredicate {
...
 
     Status update(const Field& value, const String& col_name, bool is_reverse);
 
 private:
     mutable std::shared_mutex _rwlock;
     Field _orderby_extrem {Field::Types::Null};
-    std::shared_ptr<ColumnPredicate> _predictate;
+    std::shared_ptr<ColumnPredicate> _predicate;
     TabletSchemaSPtr _tablet_schema = nullptr;
     std::unique_ptr<Arena> _predicate_arena;
```
Comment (HappenLee) [inline be/src/runtime/runtime_predicate.h:77]: Arena should be member ，not ptr

File: be/src/runtime/runtime_predicate.h [hunk 5]
```diff
@@ -52,25 +53,32 @@ class RuntimePredicate {
...
     }
 
-    std::shared_ptr<ColumnPredicate> get_predictate() {
+    std::shared_ptr<ColumnPredicate> get_predicate() {
         std::shared_lock<std::shared_mutex> rlock(_rwlock);
-        return _predictate;
+        return _predicate;
     }
 
     Status update(const Field& value, const String& col_name, bool is_reverse);
```
Comment (HappenLee) [inline be/src/runtime/runtime_predicate.h:70]: remove param of col_name

File: be/src/olap/shared_predicate.h [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+#include "common/factory_creator.h"
+#include "olap/column_predicate.h"
+#include "olap/rowset/segment_v2/bloom_filter.h"
+#include "olap/rowset/segment_v2/inverted_index_reader.h"
+#include "olap/wrapper_field.h"
+#include "vec/columns/column_dictionary.h"
+
+namespace doris {
+
```
Comment (HappenLee) [inline be/src/olap/shared_predicate.h:31]: Add comment of the class

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]

**Review by Gabriel39 — APPROVED** [review]

**Pull request merged**